java/ql/src/experimental/Security/CWE/CWE-400/LocalThreadResourceAbuse.ql

```diff
@@ -0,0 +1,111 @@
+/**
+ * @name Uncontrolled thread resource consumption from local input source
+ * @description Use user input directly to control thread sleep time could lead to performance problems
+ *              or even resource exhaustion.
+ * @kind path-problem
+ * @id java/thread-resource-abuse
+ * @problem.severity recommendation
+ * @tags security
+ *       external/cwe/cwe-400
+ */
+
+import java
+import ThreadPauseSink
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+/** The `getInitParameter` method of servlet or JSF. */
+class GetInitParameter extends Method {
+  GetInitParameter() {
+    (
+      this.getDeclaringType()
+          .getASupertype*()
+          .hasQualifiedName(["javax.servlet", "jakarta.servlet"],
+            ["FilterConfig", "Registration", "ServletConfig", "ServletContext"]) or
+      this.getDeclaringType()
+          .getASupertype*()
+          .hasQualifiedName(["javax.faces.context", "jakarta.faces.context"], "ExternalContext")
+    ) and
+    this.getName() = "getInitParameter"
+  }
+}
+
+/** An access to the `getInitParameter` method. */
+class GetInitParameterAccess extends MethodAccess {
+  GetInitParameterAccess() { this.getMethod() instanceof GetInitParameter }
+}
+
+/* Init parameter input of a Java EE web application. */
+class InitParameterInput extends LocalUserInput {
+  InitParameterInput() { this.asExpr() instanceof GetInitParameterAccess }
+}
+
+private class LessThanSanitizer extends DataFlow::BarrierGuard {
+  LessThanSanitizer() { this instanceof ComparisonExpr }
+
+  override predicate checks(Expr e, boolean branch) {
+    e = this.(ComparisonExpr).getLesserOperand() and
+    branch = true
+    or
+    e = this.(ComparisonExpr).getGreaterOperand() and
+    branch = false
+  }
+}
+
+/** Taint configuration of uncontrolled thread resource consumption from local user input. */
+class ThreadResourceAbuse extends TaintTracking::Configuration {
+  ThreadResourceAbuse() { this = "ThreadResourceAbuse" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof LocalUserInput }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof PauseThreadSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(
+      Method rm, ClassInstanceExpr ce, Argument arg, Parameter p, FieldAccess fa, int i // thread.start() invokes the run() method of thread implementation
+    |
+      rm.hasName("run") and
+      ce.getConstructedType().getSourceDeclaration() = rm.getSourceDeclaration().getDeclaringType() and
+      ce.getConstructedType().getASupertype*().hasQualifiedName("java.lang", "Runnable") and
+      ce.getArgument(i) = arg and
+      ce.getConstructor().getParameter(i) = p and
+      fa.getEnclosingCallable() = rm and
+      DataFlow::localExprFlow(p.getAnAccess(), fa.getField().getAnAssignedValue()) and
+      node1.asExpr() = arg and
+      node2.asExpr() = fa
```

Comment:
@smowton - Thanks for reviewing this PR. Sorry I don't quite get it. The constructor argument is from user input as the remote flow source, then it's assigned to a field, then the <code>run</code> method accesses it. This is the best we can get so far with the help from @Marcono1234. Please help if we can further improve it.

Reply:
After discussing this with @smowton, he had a great idea to simplify this taint step and making it less ad-hoc. We could just add a value step from the qualifier of `start()` to the instance parameter (`this`) of `run`. That way the dataflow library can handle the rest (field accesses, method calls, and so on). You could remove the existing additional taint step and add this:

```ql
private class RunnableStartToRunStep extends AdditionalValueStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    exists(MethodAccess ma, Method m |
      ma.getMethod().hasName("start") and
      ma.getQualifier().getType() = m.getDeclaringType() and
      m.getDeclaringType().getASupertype*().hasQualifiedName("java.lang", "Runnable") and
      m.hasName("run")
    |
      pred.asExpr() = ma.getQualifier() and
      succ.(DataFlow::InstanceParameterNode).getEnclosingCallable() = m
    )
  }
}
```

Take into account that `AdditionalValueStep` comes from `semmle.code.java.dataflow.FlowSteps`, and you will need to rebase `main` to use it.

Finally, if you think that the `start()` call could be hidden, e.g. in a library (`myExternalLib.runLater(someWork)` instead of `someWork.start()`) and it's better to just use the task's constructor as the `pred` node, feel free to change it.

Reply:
Thanks @atorralba and @smowton a lot for the code and the detailed explanation. I've modified the query and use the task's constructor as the <node>pred</node> node as per the suggestion.